src/plugins/es_ui_shared/static/forms/hook_form_lib/types.ts

```diff
@@ -0,0 +1,144 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { ChangeEvent, FormEvent, MouseEvent, MutableRefObject } from 'react';
+import { Subject } from './lib';
+
+export interface Form<T = FormData> {
+  readonly isSubmitted: boolean;
+  readonly isSubmitting: boolean;
+  readonly isValid: boolean;
+  readonly options: FormOptions;
+  onSubmit: (e?: FormEvent<HTMLFormElement> | MouseEvent) => Promise<{ data: T; isValid: boolean }>;
+  setFieldValue: (fieldName: string, value: FieldValue) => void;
+  setFieldErrors: (fieldName: string, errors: ValidationError[]) => void;
+  readonly __formData$: MutableRefObject<Subject<T>>;
+  __addField: (field: Field) => void;
+  __removeField: (fieldNames: string | string[]) => void;
+  __validateFields: (fieldNames?: string[]) => Promise<boolean>;
+  __getFormData: (options?: { unflatten?: boolean }) => T;
+  __updateFormDataAt: (field: string, value: unknown) => T;
+  __getFieldDefaultValue: (fieldName: string) => unknown;
+  __readFieldConfigFromSchema: (fieldName: string) => FieldConfig;
+}
+
+export interface FormSchema<T = FormData> {
+  [key: string]: FormSchemaEntry<T>;
+}
+type FormSchemaEntry<T> =
+  | FieldConfig<T>
+  | Array<FieldConfig<T>>
+  | { [key: string]: FieldConfig<T> | Array<FieldConfig<T>> | FormSchemaEntry<T> };
+
+export interface FormConfig<T = FormData> {
+  onSubmit?: (data: T, isFormValid: boolean) => void;
+  schema?: FormSchema<T>;
+  defaultValue?: Partial<T>;
+  serializer?: SerializerFunc<T>;
+  deSerializer?: SerializerFunc;
+  options?: FormOptions;
+}
+
+export interface FormOptions {
+  errorDisplayDelay: number;
+  /**
+   * Remove empty string field ("") from form data
+   */
+  stripEmptyFields: boolean;
```

Comment:
Indeed there is not a _need_ to offer it to the consumer, but it brings a lot of value to him and it does not have a great cost.

In (very) rare cases we want to send this type of object in a payload:

```js
{
  name: 'John',
  lastname: '',
  email: '',
  title: '',
}
```

We probably only want to send what has been modified by a user

```js
{
  name: 'John',
}
```

So as it is such a common case, I think it very practical that the lib offers a way to remove untouched fields to free the consumer to have to sanitize everywhere its data.

The `sanitizer` that you mention can currently be done through the `serializer` and `deserializer` functions.

Reply:
There is no concept of "editing" "creating" to the library. The only thing it says is: "do you want field that are empty strings `""` to be removed from the form data when being returned. So in our apps, we always want it, as we never do a `PATCH` request, but we replace the whole resource in ES (`POST` or `PUT`).

> Does a series of whitespace characters count as an empty string?

Yes multiple whitespace === an empty field.

> If a default value is specified for a field and the user deletes it, will the default value be re-inserted?

If a default value is provided but the user removes it (empty the field), then it will not appear in the output object. A validation might not allow the field to be empty, but that's another topic 😊